feat: support nested STRUCT and ARRAY data display in anywidget mode

shuoweil opened the pull request:
Implements flattening and expansion for complex data types in the interactive display for anywidget mode.

  Key Features:
   * Automatic Flattening: STRUCT columns are flattened into intuitive dot-notation columns (e.g., user.name).
   * Array Expansion: ARRAY columns are expanded into multiple rows with visual grouping.
   * Visual Continuity: Continuation rows for arrays are styled for better parent-row context.

verified at:
- vs code notebook: screen/3ST4m9xN9w3iqD9
- colab notebook: screen/7NG4LiTEPuAC27F

Fixes #<438181139> 🦕

shuoweil committed:
refactor(display): use CSS classes in HTML tables

shuoweil committed:
refactor(display): use CSS classes in HTML tables

shuoweil committed:
feat(display): support nested STRUCT and ARRAY data in interactive tables

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
chore: remove unreached code

shuoweil committed:
refactor: code refactor

shuoweil committed:
refactor: resue pandas struct.explode()

shuoweil committed:
refactor: revert the refactor

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-ui-improve

shuoweil committed:
test: merge notebook

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
feat: use dataclass for flatten_nested_data

shuoweil committed:
feat: Refactor HTML rendering and document JS tests

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

tswast commented on bigframes/display/_flatten.py:
```diff
@@ -0,0 +1,313 @@
+            continue
+
+        # Create one row per array element, up to max_len_in_row
+        for array_idx in range(max_len_in_row):
```
This is looping through each element of each array in Python, which is going to be even slower.

shuoweil replied:
I have completely refactored _explode_array_columns to use a vectorized approach with pandas.explode and merge. This eliminated all Python loops, including the slow inner loop you pointed out, significantly improving performance.

shuoweil committed:
Fix: Improve performance of nested data flattening

shuoweil commented on bigframes/display/_flatten.py:
```diff
@@ -0,0 +1,313 @@
+    )
+
+
+def _flatten_array_of_struct_columns(
```
Thanks for the suggestion. I've refactored the array explosion logic to use a much faster vectorized approach with pandas.explode and merge, which removes the Python loops entirely.

shuoweil committed:
Fix: Correct bug in nested data flattening

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
fix(display): fix explode on Arrow list columns in flatten_nested_data

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
perf(display): optimize nested data flattening and fix js style

shuoweil committed:
test: rerun notebook

shuoweil committed:
fix(display): add row hover effect for nested data rows

shuoweil committed:
refactor: code refactor

shuoweil committed:
refactor: improve _flatten readability and table widget styles

shuoweil committed:
docs: move implementation details from docstrings to block comments

shuoweil committed:
docs: remove redundant comments in _flatten.py

shuoweil committed:
refactor: simplify flattening logic in _flatten.py

shuoweil committed:
refactor: use mutable ColumnClassification object in _flatten.py

shuoweil committed:
fix: resolve bug in _classify_columns logic and enable functional updates

shuoweil committed:
refactor: simplify _classify_columns logic in _flatten.py

shuoweil committed:
fix: resolve NameError for ExplodeResult and formatting

shuoweil committed:
refactor(anywidget): optimize and style cleanup for flatten logic

- Replaced Python-based row explosion with optimized PyArrow computation for nested arrays.
- Cleaned up comments in  to strictly adhere to Google Python Style Guide (focused on 'why', removed redundant 'what').
- Renamed variable  to  for clarity.
- Verified changes with Python unit tests and JavaScript frontend tests.

shuoweil committed:
refactor(anywidget): optimize array flattening using pyarrow

shuoweil committed:
test: rerun notebook

shuoweil committed:
refactor: remove nested loop

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
Merge main to shuowei-anywidget-nested-strcut-array

shuoweil committed:
test: rerun notebook to verify the merge

tswast commented on bigframes/display/_flatten.py:
```diff
@@ -0,0 +1,528 @@
+                    if pa.types.is_struct(pa_type.value_type)
+                    else "array"
+                )
+        return "clear"
```
These magic strings worry me. Could you create an enum for category, instead?

https://docs.python.org/3/library/enum.html

tswast commented on bigframes/display/_flatten.py:
```diff
@@ -0,0 +1,528 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class ColumnClassification:
```
Please put a leading `_` in front of class names that aren't intended to be used outside of this module.

tswast commented on bigframes/display/html.py:
```diff
@@ -117,6 +130,10 @@ def render_col_header(col):
     dataframe: pd.DataFrame,
+    row_labels: list[str] | None,
+    continuation_rows: set[int] | None,
+    clear_on_continuation: list[str],
```
Same here, add some more explanation to the docstrings. To keep it shorter, you could reference bigframes/display/_flatten.py so that folks can look there for the complete explanation.

shuoweil committed:
Merge commit '798af4a30' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
refactor: replace magic strings for col categories with a private Enum

shuoweil committed:
refactor: replace magic strings for col categories with a private Enum

shuoweil committed:
test: rerun notebook

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
docs: rerun test

shuoweil committed:
test: update year

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-nested-strcut-array
